crates/agglayer-contracts/src/lib.rs

```diff
@@ -145,35 +157,146 @@ pub enum L1RpcError {
     CacheLockPoisoned,
 }
 
+/// Finds the block number where a contract was deployed using binary search.
+/// This is used to optimize event queries by starting from the deployment
+/// block.
+async fn find_contract_deployment_block_number<RpcProvider>(
+    rpc: &RpcProvider,
+    address: Address,
+) -> Result<Option<u64>, L1RpcError>
+where
+    RpcProvider: alloy::providers::Provider,
+{
+    use alloy::eips::{BlockId, BlockNumberOrTag};
+
+    // Get the latest block number as the upper bound for search
+    let latest_block = rpc.get_block_number().await.map_err(|e| {
+        L1RpcError::FailedToQueryEvents(format!("Failed to get latest block number: {}", e))
+    })?;
+
+    debug!("Finding contract deployment block for address: {address}, latest_block {latest_block}");
+
+    let mut hi = latest_block;
+    let mut lo = 0u64;
+
+    // Quick exit: if no code at latest block, contract was never deployed
+    // (Note: this will treat self-destructed contracts as "not found")
+    let latest_code = rpc
+        .get_code_at(address)
+        .block_id(BlockId::Number(BlockNumberOrTag::Number(hi)))
+        .await
+        .map_err(|e| {
+            L1RpcError::FailedToQueryEvents(format!("Failed to get code at latest block: {}", e))
+        })?;
+
+    if latest_code.is_empty() {
+        debug!("No code found at latest block for address: {}", address);
+        return Ok(None);
+    }
+
+    debug!(
+        "Contract has code at latest block, starting binary search from block 0 to {}",
+        hi
+    );
+
+    // Binary search for the first block where code is present
+    let mut answer: Option<u64> = None;
+    let mut iterations = 0;
+    const MAX_ITERATIONS: u32 = 256; // Safety limit to prevent infinite loops
+
+    while lo <= hi && iterations < MAX_ITERATIONS {
+        iterations += 1;
+        let mid = lo + ((hi - lo) / 2);
+
+        debug!(
+            "Binary search iteration {}: checking block {} (range: {} to {})",
+            iterations, mid, lo, hi
+        );
+
+        // Query bytecode at specific block height
+        let code_at_mid = rpc
+            .get_code_at(address)
+            .block_id(BlockId::Number(BlockNumberOrTag::Number(mid)))
+            .await
+            .map_err(|e| {
+                L1RpcError::FailedToQueryEvents(format!(
+                    "Failed to get code at block {}: {}",
+                    mid, e
+                ))
+            })?;
+
+        if code_at_mid.is_empty() {
+            // Not deployed yet at `mid` -> search higher
+            debug!("No code at block {}, searching higher", mid);
+            lo = mid.saturating_add(1);
+        } else {
+            // Code exists at `mid` -> remember and search lower
+            debug!("Code found at block {}, searching lower", mid);
+            answer = Some(mid);
```

Comment:
I don't think this is correct if the loop exits early. To get a safe conservative starting point for log scanning, the answer should be set at point where the code does not yet exist, so in the other branch of the if-else to the updated `lo`.

Reply:
I have removed this function, querying from `Earliest`

Reply:
Not really, only in rather synthetic scenarios. E.g. when someone decides to limit the iteration count to some smaller number (say 18), where the loop exits prematurely and there still is a gap between lo and hi. The contract is deployed at some point x, where lo <= x <= mid, and mid is returned. When scanning the logs starting from mid, the logs between contract deployment x and mid will be missed.

So more a matter of robustness in terms of parameter tuning and code consistency. The more invariants can you spell out, the easier it is to reason about the code in general. (In fact, I initially assumed it to be true that the answer always contains either `None` or a conservative bound for start of log scanning.) But I might have missed a different corner case...